swarm/core.py

```diff
@@ -80,3 +85,128 @@ def swarm_train(bee_trainer, num_swarm=50, seed=None, fields=None):
+        dync = {k: v for k, v in zip(kwargs, dynamic)}
+        yield merge_dicts(static, dync)
+
+
+def make_combo_paramsets(static: Dict[str, Any], *combo: Dict[str, Sequence], **kwargs: Sequence):
+    """
+    A function for building a paramset for hive-style training. For a bee, we may have a series
+    of parameters we're interested in and want to generate data across a sweep of various params
+
+    These dictionaries are used by the training procedure to configure the bee as passed in
+    via **kwargs
+
+    Args:
+        static: dictionary of static params that do not change
+        *combo: A way to configure the params to sweep across.
+            Formed by {"param": [p1, p2, ... pn], "other": [o1, o2, .., on]}
+        **kwargs: A convenience method for single vars instead of doing
+        {"lr": (0.001, 0.002, 0.004)} you can simply do
+        lr = (0.001, 0.002, 0.004) in the function call.
+
+    Yields:
+        Dict[str, Any] - the params to feed into a bee as **kwargs
+
+    Examples:
+        See hive_experiment.py
+        >>> pm = make_combo_paramsets({"x": 3}, \
+            {"lr": (0.001, 0.004), "momentum": (.9, .95)}, \
+            activation=(torch.nn.Sigmoid, torch.nn.ReLU))
+        >>> for p in pm:
+        ...    print(p)
+        {"x": 3, "lr": 0.001, "momentum": .9, "activation":torch.nn.Sigmoid}
+        {"x": 3, "lr": 0.004, "momentum": .95, "activation":torch.nn.Sigmoid}
+        {"x": 3, "lr": 0.001, "momentum": .9, "activation":torch.nn.ReLU}
+        {"x": 3, "lr": 0.004, "momentum": .95, "activation":torch.nn.ReLU}
+
+        Note that lr and momentum are paired while activation is independent.
+    """
+    # remove empty dicts
+    combo = [c for c in combo if c]
+    if not combo:
+        yield from make_sweep_paramsets(static, **kwargs)
+    else:
+        for dyn_combo in itertools.product(*(_dict_slicer(combodict) for combodict in combo)):
+            newst = merge_dicts(static, *dyn_combo)
+            yield from make_sweep_paramsets(newst, **kwargs)
+
+
+def hive_trainer(bee: Callable, param_list: Iterable[Dict[str, Any]],
+                 num_swarm: int = 50, seed: int = None, fields: str = None):
+    """
+    This extends swarm_train to do a sweep across parameter_list.
+
+    Args:
+        bee: Callable
+        param_list: An iterable of kwargs to be passed to the bee
+
+    Returns:
+        List[Dict[str, Any]] - Tidy-esque data containing both the parameter and the bee result
+        Will look like [{"lr": 0.01, "momentum": 0.95, "ypred": np.array, "loss", np.array},
+                         "lr": 0.01, "momentum": 0.9, "ypred": np.array, "loss", np.array]]
+        You can call pd.DataFrame on this result to turn it into a dataframe.
+
+    """
+    ret = []
+    for param in param_list:
+        res = swarm_train(
+            bee, param,
+            num_swarm, seed, fields)
+        ret.append({**param, **res})
+    return ret
```

Comment:
This all looks very neat.  And great documentation. It is readable. I'll have to have a play with it to get a sense of how it works. I suspect that the python version of slightly nested tidy-esque data might need one step more before transferred to R. I feel a little lost at the moment between persisting with pushing my suggested changes separately, and just waiting for you to merge all of these in. Can discuss today I guess.